sw/host/opentitanlib/src/util/status.rs

```diff
@@ -0,0 +1,255 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// This utility handles the various aspect of the status_t type used in
+// the device code.
+
+use anyhow::{bail, Context, Result};
+use bindgen::status::{ot_status_create_record_t, status_create, status_err, status_extract};
+
+use object::{Object, ObjectSection};
+
+use num_enum::TryFromPrimitive;
+
+use std::convert::TryFrom;
+use std::ffi::CString;
+use std::path::PathBuf;
+
+pub use bindgen::status::absl_status_t as RawStatusCode;
+pub use bindgen::status::status_t as RawStatus;
+
+// FIXME: is there a better way of doing this? bindgen does not
+// support adding custom derive to a type.
+#[derive(Debug, serde::Serialize, serde::Deserialize, TryFromPrimitive, PartialEq, Eq)]
+#[repr(u32)]
+pub enum StatusCode {
+    Ok = bindgen::status::absl_status_code_kOk,
+    Cancelled = bindgen::status::absl_status_code_kCancelled,
+    Unknown = bindgen::status::absl_status_code_kUnknown,
+    InvalidArgument = bindgen::status::absl_status_code_kInvalidArgument,
+    DeadlineExceeded = bindgen::status::absl_status_code_kDeadlineExceeded,
+    NotFound = bindgen::status::absl_status_code_kNotFound,
+    AlreadyExists = bindgen::status::absl_status_code_kAlreadyExists,
+    PermissionDenied = bindgen::status::absl_status_code_kPermissionDenied,
+    ResourceExhausted = bindgen::status::absl_status_code_kResourceExhausted,
+    FailedPrecondition = bindgen::status::absl_status_code_kFailedPrecondition,
+    Aborted = bindgen::status::absl_status_code_kAborted,
+    OutOfRange = bindgen::status::absl_status_code_kOutOfRange,
+    Unimplemented = bindgen::status::absl_status_code_kUnimplemented,
+    Internal = bindgen::status::absl_status_code_kInternal,
+    Unavailable = bindgen::status::absl_status_code_kUnavailable,
+    DataLoss = bindgen::status::absl_status_code_kDataLoss,
+    Unauthenticated = bindgen::status::absl_status_code_kUnauthenticated,
+}
+
+// Safe abstraction over status_create
+fn status_create_safe(code: StatusCode, mod_id: u32, file: String, arg: i32) -> RawStatus {
+    // We do not expect an error since it is a valid String.
+    let file = CString::new(file).expect("CString::new failed");
+    unsafe {
+        // Safety: the function expects a valid readonly C-string which is exactly what
+        // CString:as_ptr() provides.
+        status_create(code as u32, mod_id, file.as_ptr(), arg)
+    }
+    // Note: file is dropped here so the C-string pointer is valid accross the function call.
+}
+
+// Convert an array of i8 to a string. This function will stop at first 0 (or at the
+// end of the array if it contains no zero).
+fn c_string_to_string(array: &[i8]) -> Result<String> {
+    // It might contain a lot of zeroes at the end which CString does not like and that we need to remove.
+    let s = array
```

Comment:
If we upgrade rustc to 1.69 we have [from_bytes_until_nul](https://doc.rust-lang.org/stable/std/ffi/struct.CStr.html#method.from_bytes_until_nul)

Reply:
I don't think the code will be terribly simpler but I can switch to that.